src/passes/MultiMemoryLowering.cpp

```diff
@@ -0,0 +1,394 @@
+            builder.makeGlobalGet(offsetGlobalName, pointerType),
+            getPageDelta()),
+          // source
+          builder.makeGlobalGet(offsetGlobalName, pointerType),
+          // size
+          builder.makeBinary(
+            Abstract::getBinary(pointerType, Abstract::Sub),
+            builder.makeMemorySize(combinedMemory, memoryInfo),
+            builder.makeGlobalGet(offsetGlobalName, pointerType)),
+          combinedMemory,
+          combinedMemory));
+    }
+
+    // Adjust the offsets of the globals impacted by the memory.grow call
+    for (Index i = memIdx; i < offsetGlobalNames.size(); i++) {
+      auto& offsetGlobalName = offsetGlobalNames[i];
+      functionBody = builder.blockify(
+        functionBody,
+        builder.makeGlobalSet(
+          offsetGlobalName,
+          builder.makeBinary(
+            Abstract::getBinary(pointerType, Abstract::Add),
+            builder.makeGlobalGet(offsetGlobalName, pointerType),
+            getPageDelta())));
+    }
+
+    functionBody = builder.blockify(
+      functionBody, builder.makeLocalGet(returnLocal, pointerType));
+
+    function->body = functionBody;
+    return function;
+  }
+
+  // This function replaces memory.size instructions with a function that can
+  // return the size of each memory as if each was discrete and separate.
+  std::unique_ptr<Function> memorySize(Index memIdx) {
+    Builder builder(*wasm);
+    Name name = "custom_memory_size_" + std::to_string(memIdx);
+    Name functionName = Names::getValidFunctionName(*wasm, name);
+    auto function = Builder::makeFunction(
+      functionName, Signature(Type::none, pointerType), {});
+    Expression* functionBody;
+
+    // offsetGlobalNames does not keep track of a global for the offset of
+    // wasm->memories[0] because it's always 0. As a result, the below
+    // calculations that involve offsetGlobalNames are intrinsically "offset".
+    // Thus, offsetGlobalNames[0] is the offset for wasm->memories[1] and
+    // the size of wasm->memories[0].
+    if (memIdx == 0) {
+      auto offsetGlobalName = offsetGlobalNames[0];
+      functionBody = builder.makeReturn(
+        builder.makeGlobalGet(offsetGlobalName, pointerType));
+    } else if (isLastMemory(memIdx)) {
+      auto offsetGlobalName = getOffsetGlobal(memIdx);
+      functionBody = builder.makeReturn(builder.makeBinary(
+        Abstract::getBinary(pointerType, Abstract::Sub),
+        builder.makeMemorySize(combinedMemory, memoryInfo),
+        builder.makeGlobalGet(offsetGlobalName, pointerType)));
+    } else {
+      auto offsetGlobalName = getOffsetGlobal(memIdx);
+      auto nextOffsetGlobalName = getOffsetGlobal(memIdx + 1);
+      functionBody = builder.makeReturn(builder.makeBinary(
+        Abstract::getBinary(pointerType, Abstract::Sub),
+        builder.makeGlobalGet(nextOffsetGlobalName, pointerType),
+        builder.makeGlobalGet(offsetGlobalName, pointerType)));
+    }
```

Comment:
There are three cases here, but the test only has two memories so I think we are missing test coverage for one of them.